### ISSUE TITLE
Order of Points in a Structure's 'contour' Field Impacts Metrics Results

### ISSUE THREAD

**Issue by dgourd** [cerr/CERR#209]
Hello,

I have CT data that has already been segmented in another software, [Amira by ThermoFisher](https://www.fei.com/software/amira-avizo/), that I am trying to analyze in CERR.  I am able to import the DICOMs into CERR and wrote this MATLAB script ([amLabeltoCERR.zip](https://github.com/cerr/CERR/files/2543589/amLabeltoCERR.zip)) to convert the Amira ROIs into a MATLAB structure and add them to the planC object generated by CERR.  However, when testing this script, I noticed that the ordering of the points in the contours affects the results of the intensity metrics output by CERR for the exact same ROIs.

In my testing, I created the same ROI with 3 different point orders: the natural order generated by the FOR loops in my script, randomly ordered points using MATLAB's [randperm function](https://www.mathworks.com/help/matlab/ref/randperm.html), and sorted points using MATLAB's [sortrows function](https://www.mathworks.com/help/matlab/ref/sortrows.html).  Despite having the exact same voxels included in each ROI, the different ordering of the points gave me the following results:
### FOR loop order:
![abb20181002_rat37_labeled](https://user-images.githubusercontent.com/8401152/47932756-89a18e00-dea0-11e8-88b6-7a6f21d352b8.PNG)
### RANDPERM order:
![abb20181002_rat37_labeled_randperm](https://user-images.githubusercontent.com/8401152/47932757-89a18e00-dea0-11e8-95d5-7a1df671e3ac.PNG)
### SORTROWS order:
![abb20181002_rat37_labeled_sortrows](https://user-images.githubusercontent.com/8401152/47932758-89a18e00-dea0-11e8-8928-cfee15e8e238.PNG)

For some reason, the mean CT intensity and total volume vary depending on the sorting mechanism.  When visually inspecting the structures in the CERR viewer, the ROIs seem to be imported properly and indeed include the same points.  I looked at the order of points in structures I created within CERR and noticed that each set is enclosed, with the first and last rows of points being the same.  However, whether I did or did not enclose my imported ROIs had no impact on the metrics.

I did notice that the CERR generated ROIs used some sort of circular sorting mechanism.  Despite this, I feel as though the order of the points shouldn't impact the results of metrics such as mean intensity and volume and that this might be unintended.  Is there a specific sorting of points that CERR assumes when running it's metrics?

Thank you,
dgourd

CERR GitHub Hash: 9f300b15d03facd1fe31b1e339d3031ce0a9b380
CERR pull Date: 26-Oct-2018 10:02:38
MATLAB Version: R2017b
Amira Version: 6.5

**Comment by adityaapte** [comment]
CERR creates raster segments from the contour polygonal points. In order to
check whether the same voxels get included for all the structures, use the
following in CERR command box:
mask structureNum. For example, mask 1 to look at rastersegments for
structure number 1 and so on. Can you share your anonymized CERR dataset?

**Comment by dgourd** [comment]
Running the mask command on the unsorted structure generated no output, neither in the MATLAB console nor in the UI.  However, running it on the sorted structure gave me a mask that didn't match the ROI.  Here is a comparison of a rat lung ROI from the raster mask vs the display when enabling 'Contouring' in the 'Structures' menu:
![rat_lung_raster_mask](https://user-images.githubusercontent.com/8401152/48027449-6c71f700-e10e-11e8-9d1f-abcf0975ba1f.PNG)
![rat_lung_contouring](https://user-images.githubusercontent.com/8401152/48027451-6ed45100-e10e-11e8-8ab9-a76085c46b38.PNG)
The raster doesn't seem to handle the bottom right corner of both segments.  Any suggestions on what I can do to make my imported structures play nice with your raster segmentation?

The only dataset I can provide is a CT of an Air/Water HU calibration phantom.  [Here is a .zip](https://northwestern.box.com/s/ie0wlccl9547621vtudrrfmy42s093xa) of the CERR dataset with an Amira ROI label so you can test the script.

**Comment by dgourd** [comment]
@adityaapte thank you for the mask command.  That really helped me debug this issue on my own.  I found that instead of storing all the voxels in an ROI in the points matrix, storing only the boundary voxels in an enclosed, ordered list resulted in generating the intended rastersegments for each ROI.  I used MATLAB's [bwboundaries function](https://www.mathworks.com/help/images/ref/bwboundaries.html) as a convenient way to generate these boundary point matrices.
